daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/svalue/Ordering.scala

```diff
@@ -0,0 +1,195 @@
+      Ast.BTOptional,
+      Ast.BTList,
+      Ast.BTTextMap,
+      Ast.BTGenMap,
+      Ast.BTAny,
+      Ast.BTTypeRep,
+      Ast.BTUpdate,
+      Ast.BTScenario
+    ).zipWithIndex.toMap
+
+  private def typeRank(typ: Ast.Type): Int =
+    typ match {
+      case Ast.TBuiltin(_) => 0
+      case Ast.TTyCon(_) => 1
+      case Ast.TNat(_) => 2
+      case Ast.TStruct(_) => 3
+      case Ast.TApp(_, _) => 4
+      case Ast.TVar(_) | Ast.TForall(_, _) | Ast.TSynApp(_, _) =>
+        throw new IllegalArgumentException(s"cannot compare types $typ")
+    }
+
+  @tailrec
+  // Any two ground types (types without variable nor quantifiers) can be compared.
+  private[this] def compareType(x: Int, stack0: FrontStack[(Ast.Type, Ast.Type)]): Int =
+    stack0 match {
+      case FrontStack() =>
+        x
+      case FrontStackCons(tuple, stack) =>
+        if (x != 0) x
+        else
+          tuple match {
+            case (Ast.TBuiltin(b1), Ast.TBuiltin(b2)) =>
+              compareType(builtinTypeIdx(b1) compareTo builtinTypeIdx(b2), stack)
+            case (Ast.TTyCon(con1), Ast.TTyCon(con2)) =>
+              compareType(compareIdentifier(con1, con2), stack)
+            case (Ast.TNat(n1), Ast.TNat(n2)) =>
+              compareType(n1 compareTo n2, stack)
+            case (Ast.TStruct(fields1), Ast.TStruct(fields2)) =>
+              compareType(
+                math.Ordering
+                  .Iterable[String]
+                  .compare(fields1.toSeq.map(_._1), fields2.toSeq.map(_._1)),
+                zipAndPush(fields1.iterator.map(_._2), fields2.iterator.map(_._2), stack)
+              )
+            case (Ast.TApp(t11, t12), Ast.TApp(t21, t22)) =>
+              compareType(0, (t11, t21) +: (t12, t22) +: stack)
+            case (t1, t2) =>
+              typeRank(t1) compareTo typeRank(t2)
+          }
+    }
+
+  // undefined if `typ1` or `typ2` contains `TVar`, `TForAll`, or `TSynApp`.
+  private def compareType(typ1: Ast.Type, typ2: Ast.Type): Int =
+    compareType(0, FrontStack((typ1, typ2)))
+
+  private def compareText(text1: String, text2: String): Int =
+    Utf8.Ordering.compare(text1, text2)
+
+  private def isHexa(s: String): Boolean =
+    s.forall(x => '0' < x && x < '9' || 'a' < x && x < 'f')
+
+  @inline
+  private val underlyingCidDiscriminatorLength = 65
+
+  private def compareCid(cid1: Ref.ContractIdString, cid2: Ref.ContractIdString): Int = {
+    val lim = cid1.length min cid2.length
+
+    @tailrec
+    def lp(i: Int): Int =
+      if (i < lim) {
+        val x = cid1(i)
+        val y = cid2(i)
+        if (x != y) x compareTo y else lp(i + 1)
+      } else {
+        if (lim == underlyingCidDiscriminatorLength && (//
+          cid1.length != underlyingCidDiscriminatorLength && isHexa(cid2) || //
+          cid2.length != underlyingCidDiscriminatorLength && isHexa(cid1)))
+          throw new IllegalArgumentException(
+            "Try to a fresh contract id with  conflicting discriminators")
```

Comment:
I think there are one or two missing words in this error message.

Reply:
fixed